pkg/usagetracker/tracker_test.go

```diff
@@ -3,15 +3,355 @@
+	})
+
+	t.Run("new instance takes partitions, old shut downs them", func(t *testing.T) {
+		t.Parallel()
+
+		// lostPartitionsShutdownGracePeriod should be long enough for this test not to be flaky,
+		// but also short enough to keep the test fast
+		const lostPartitionsShutdownGracePeriod = time.Second
+		configure := func(cfg *Config) {
+			cfg.LostPartitionsShutdownGracePeriod = lostPartitionsShutdownGracePeriod
+			cfg.MaxPartitionsToCreatePerReconcile = testPartitionsCount
+		}
+
+		ikv, pkv, cluster := prepareKVStoreAndKafkaMocks(t)
+		a0 := newTestUsageTracker(t, 0, "zone-a", ikv, pkv, cluster, configure)
+		b0 := newTestUsageTracker(t, 0, "zone-b", ikv, pkv, cluster, configure)
+		trackers := map[string]*UsageTracker{
+			"a0": a0,
+			"b0": b0,
+		}
+		waitUntilAllTrackersSeeAllInstancesInTheirZones(t, trackers)
+
+		// Create trackers and check that they're ready.
+		require.NoError(t, a0.reconcilePartitions(context.Background()))
+		require.NoError(t, b0.reconcilePartitions(context.Background()))
+		requireAllTrackersReady(t, trackers)
+
+		// Start zone-a-1.
+		a1 := newTestUsageTracker(t, 1, "zone-a", ikv, pkv, cluster, configure)
+		trackers["a1"] = a1
+
+		waitUntilAllTrackersSeeAllInstancesInTheirZones(t, trackers)
+
+		// Reconcile both, they will create all the needed partitions.
+		require.NoError(t, a0.reconcilePartitions(context.Background()))
+		require.NoError(t, a1.reconcilePartitions(context.Background()))
+
+		// All trackers should be ready.
+		requireAllTrackersReady(t, trackers)
+
+		// zone-a-0 should have 16 partitions, zone-a-1 should have 8, this is deterministic at this point.
+		a0.partitionsMtx.RLock()
+		require.Len(t, a0.partitions, 16)
+		a0.partitionsMtx.RUnlock()
+		a1.partitionsMtx.RLock()
+		require.Len(t, a1.partitions, 8)
+		a1.partitionsMtx.RUnlock()
+
+		// Check eventually because partitionRing update might be delayed.
+		partitionRing := a0.partitionRing
+		require.EventuallyWithT(t, func(t *assert.CollectT) {
+			// First half still has 2 owners.
+			for partitionID := int32(0); partitionID < testPartitionsCount/2; partitionID++ {
+				require.Len(t, partitionRing.PartitionRing().MultiPartitionOwnerIDs(partitionID, nil), 2, "Partition %d should have 2 owners.", partitionID)
+			}
+			// Second half of partitions should have 3 owners now (zone-a-0, zone-a-1, zone-b-1)
+			for partitionID := int32(testPartitionsCount / 2); partitionID < testPartitionsCount; partitionID++ {
+				require.Len(t, partitionRing.PartitionRing().MultiPartitionOwnerIDs(partitionID, nil), 3, "Partition %d should have 2 owners.", partitionID)
+			}
+		}, time.Second, 100*time.Millisecond)
+
+		// Reconcile zone-a-0 once more after ring is updated to make sure that all the lost partitions with their times.
+		require.NoError(t, a0.reconcilePartitions(context.Background()))
+
+		// Wait until old partitions are lost.
+		time.Sleep(lostPartitionsShutdownGracePeriod)
```

Comment:
Should be lostPartitionsShutdownGracePeriod + a small delta to reduce flakyness risks?

Reply:
No need really, you can't run the code in zero nanoseconds, and since usage-tracker stores the lost time as the actual time with wall clock data, it will always be higher.